tests/functional/api/test_errors.py

```diff
@@ -0,0 +1,195 @@
+# -*- coding: utf-8 -*-
+"""
+Functional tests for client API errors
+
+Uses create-group endpoint as it can raise all relevant client error codes.
+"""
+
+from __future__ import unicode_literals
+
+import pytest
+import base64
+
+
+from h.models.auth_client import GrantType
+
+native_str = str
+
+
+@pytest.mark.functional
+class Test400ErrorsUsingCreateGroup(object):
+    """Creating a group can raise all of the client errors we want to test"""
+
+    def test_it_returns_http_400_with_invalid_payload(self, app, token_auth_header):
+        group = {}
+
+        res = app.post_json(
+            "/api/groups", group, headers=token_auth_header, expect_errors=True
+        )
+
+        assert res.status_code == 400
+
+    def test_it_returns_a_formatted_reason_with_invalid_payload(
+        self, app, token_auth_header
+    ):
+        group = {}
+
+        res = app.post_json(
+            "/api/groups", group, headers=token_auth_header, expect_errors=True
+        )
+
+        assert res.json["reason"] == "u'name' is a required property"
+
+    def test_it_returns_http_400_if_groupid_set_on_default_authority(
+        self, app, token_auth_header
+    ):
+        group = {"name": "My Group", "groupid": "3434kjkjk"}
+        res = app.post_json(
+            "/api/groups", group, headers=token_auth_header, expect_errors=True
+        )
+
+        assert res.status_code == 400
+
+    def test_it_returns_formatted_reason_if_groupid_set_on_default_authority(
+        self, app, token_auth_header
+    ):
+        # FIXME: The `reason` is double-escaped
+        group = {"name": "My Group", "groupid": "3434kjkjk"}
+        res = app.post_json(
+            "/api/groups", group, headers=token_auth_header, expect_errors=True
+        )
+
+        assert (
+            res.json["reason"]
+            == "groupid: u'3434kjkjk' does not match u\"^group:([a-zA-Z0-9._\\\\-+!~*()']{1,1024})@(.*)$\""
+        )
+
+
+@pytest.mark.functional
+class Test403ErrorsUsingCreateGroup(object):
+    # FIXME: API currently converts all 403s to 404s
```

Comment:
I wonder if this FIXME belongs in a comment or docstring on the xfail test below?